negation-types.md

```diff
@@ -0,0 +1,555 @@
+assignable to) `~U` if `T` and `U` are disjoint (their intersection is `Never`). If `T`
+is a dynamic type, it is assignable to the fully static type `~U` if any materialization
+of `T` is disjoint from `U`. Equivalently, we can check whether `Bottom[T]` is disjoint
+from `U`; if `Bottom[T]` is not disjoint from `U`, then no other materialization can be
+disjoint from `U`, since all materializations are supertypes of `Bottom[T]`.
+
+If `U` is a dynamic type, then `T` is assignable to `~U` if the intersection of
+`Bottom[T]` and `Bottom[U]` is uninhabited.
+
+### Attributes on negation types
+
+Given a value `x: T & ~U`, how should a type checker infer the type of an expression
+like `x.attr`? In most cases, once the intersection has been fully simplified, the
+negative part of the intersection can simply be ignored. A type like `int & ~Literal[1]`
+has the same attributes as plain `int`; `object & ~str` does not differ from `object`.
+It is important, of course, that the intersection first gets simplified.
+`(int | str) & ~str` does not have the same attributes as `int`.
+
+It gets more complicated with structural types in the negative case. Consider this
+program:
+
+```python
+from typing import Protocol, ReadOnly
+
+class HasIntStr(Protocol):
+    x: ReadOnly[int | str]
+
+class HasStr(Protocol):
+    x: ReadOnly[str]
+
+obj: HasIntStr & ~HasStr
+reveal_type(obj.x)
+```
+
+(Assuming that `ReadOnly` is accepted on protocol attributes, as proposed by
+[PEP 767](https://peps.python.org/pep-0767).)
+
+Naively, we might think that this should reveal `int`: `obj.x` looks like it should be
+`(int | str) & ~str = int`. But there could be instances of `HasIntStr` that are not
+subtypes of `HasStr`, yet might return either an `int` or a `str` when the attribute is
+accessed.
+
+```python
+from typing import final
+
+@final
+class HasIntStrSneaky:
+    @property
+    def x(self) -> int | str:
+        return 1 if random() > 0.5 else "1"
+
+    @x.setter
+    def x(self, obj: int | str) -> None:
+        pass
+```
+
+Indeed, even without such a pathological implementation, since the `x` attribute could
+be mutable on implementations of `HasIntStr`, the type of the `x` attribute could change
+at any time from an `int` to a `str` and vice versa.
+
+Are there any variations of the `HasIntStr` + `HasStr` intersection where we _can_
+narrow the attribute type? It seems to me that we can only narrow if every value that is
+a member of the `HasIntStr` equivalent type must have either an `int` or a `str`
+attribute, not both.
+
+That is the case, for example, with the following classes:
+
+```python
+from typing import NamedTuple, Final, final
+from dataclasses import dataclass
+
+class HasIntStrNamedTuple(NamedTuple):
+    x: int | str
+
+class HasIntStrFinal:
+    x: Final[int | str]
+
+    def __init__(self, x: int | str) -> None:
+        self.x = x
+
+@final
+@dataclass(frozen=True)
+class HasIntStrFrozen:
+    x: int | str
+```
```

Comment:
am I correct in saying then that even for these examples, the narrowing of the attribute type is only possible here because `int` and `str` are disjoint: an object can inhabit one or the other type, but never both simultaneously?

Reply:
I don't think that matters. We can narrow to `A | B & ~B` only if the conditions listed further down hold; whether that simplifies to `A` or not is sort of an independent question.